### DISCUSSION TITLE
What's the best way to split the application logic into multiple classes/files ?

### DISCUSSION THREAD

**Discussion by bog-dan-ro** [slint-ui/slint#3005] [General]
Hello there,

In Qt (sorry for keep using it as an example), I could split the application logic into multiple classes:
 -  for QWidget `.ui` apps, it was easy, as you created all the views by inheriting the apropiate QClass, called the setup to create all the ui elements and all the UI signals/slots where isolated in that class.
 - for QML apps,  using `QML_ELEMENT` I could expose any C++ stuff to QML world easily and there I could used them independently.

I checked all slint C++ examples and, to my knowledge, it seems the only way to interact with the UI elements is through main window properties &  callbacks, which for a simple application might be okay, but for apps which have lots of "screens" with  many UI controls on each screen won't work. Even for a simple application like `Printer Demo` example, you'll endup with a few dozens of properties & callbaks in one place...
Is there anything that I'm missing? If yes, can you please change `Printer Demo`, and add a class which does something for each view? 
E.g. for :
- settings
- printing
- scanning
- copy
- usb
- etc. 

It's probably related to https://github.com/slint-ui/slint/issues/154

**Comment by ogoffart** [comment]
You can move the code that initialize the properties and callback for a given page in a different file.

If, for example you have a global for a given part of the UI, you can have this funciton in a separate file. Or maybe even as a member of a C++ class 

```c++
void init_foobar(const Foobar &) { ... }
```

And you can call that from the UI code as

```c++
init_foobar(app->global<Foobar>());
```

**Reply by romangg** [reply]
How would this work together with C++ classes or other data structs being created lazily when switching views?

Overall it would be nice to have a working official example/demo showcasing best practice for questions like this. Because I also noticed that the currently available examples are very minimal and don't demonstrate how the UI code in larger applications should be structured when aspects on the backend like data modularity and dependency factor in.

**Comment by coolerputt01** [comment]
I have also experienced this I am currently looking into _conditionally rendering pages_ using a states system while still using the `MainWindow ` as a layout system of the application.


`if app_state == "start" : start_page_initialise :=  StartPage {}`

This doesn't tho, I am getting` Parse Error`

**Reply by coolerputt01** [reply]
I later settled in using ‘visible’ property for each pages cos I can’t just initialize components that easily in Slint
